src/bootstrap/toolstate.rs

```diff
@@ -180,11 +203,24 @@ impl Step for ToolStateCheck {
                     did_error = true;
                     eprintln!("error: Tool `{}` should be test-pass but is {}", tool, state);
                 } else if in_beta_week {
-                    did_error = true;
-                    eprintln!(
-                        "error: Tool `{}` should be test-pass but is {} during beta week.",
-                        tool, state
-                    );
+                    let old_state = old_toolstate
+                        .iter()
+                        .find(|ts| ts.tool == *tool)
+                        .expect("latest.json missing tool")
+                        .state();
+                    if state < old_state {
+                        did_error = true;
+                        eprintln!(
+                            "error: Tool `{}` has regressed from {} to {} during beta week.",
+                            tool, old_state, state
+                        );
```

Comment:
Reading this code, something very odd is going on here... there are two places where the code does "if regressed and in_beta_weak, then fail": it does that here, and it does that again and seemingly independently in `change_toolstate`.

What is going on here? Seems like duplication of the same logic, which usually shouldn't happen.

Reply:
@ehuss also see https://github.com/rust-lang/rust/pull/69693. Feel free to just include that commit in your PR, then I will close mine.

Reply:
Posted #69705 with the fix.